tests/console/zdup_prepare_leap_sle.pm

```diff
@@ -0,0 +1,103 @@
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.  This file is offered as-is,
+# without any warranty.
+
+# Summary: prepare migration from Leap to SLE
+#   Documentation for 15:
+#   https://www.suse.com/de-de/documentation/sles-15/book_sle_upgrade/data/sec_upgrade-online_opensuse_to_sle.html
+#   12 is not supported and inoffical
+# Maintainer: lemon.li <leli@suse.com>
+
+use base "installbasetest";
+use registration;
+use testapi;
+use utils qw(zypper_call);
+use strict;
+use warnings;
+
+#   Function: judge base system is which version of leap
+#   parameters:
+#   $version : expected leap version
+#   return : 1 if ture
+sub is_leap_base {
+    my $version = shift;
+    return (get_var('HDDVERSION', '') =~ /opensuse-$version/);
+}
+
+sub run {
+    my $reg_code   = get_required_var('SCC_REGCODE');
+    my $scc_url    = get_required_var('SCC_URL');
+    my $scc_addons = get_var('SCC_ADDONS', '');
+
+    # so if we are migrating from Leap to SLE we just make SUSEConnect believe
+    # we are on SLE
+    my $hddversion = get_var('HDDVERSION', '');
+    my $version    = get_var('VERSION');
+    $version =~ s/-SP/./;    # 12-SP3 -> 12.3
+    my $product .= "SLES/$version/" . get_var('ARCH');
+    my $arch = get_var('ARCH');
+
+    select_console 'root-console';
+
+    # make sure we have SUSEConnect and the right build key
+    zypper_call "in SUSEConnect";
+    # Remove packages that produce file conflicts during the migration.
+    script_run "rpm -e --nodeps yast2-qt-branding-openSUSE";
+    # suse build key conflicts with openSUSE-build-key. We can't uninstall
+    # openSUSE-build-key though as patterns need it :(
+    # should be only needed on 12
+    if (is_leap_base('42')) {
+        zypper_call "download suse-build-key";
+        assert_script_run "rpm -Uvh --force --nodeps /var/cache/zypp/packages/*/noarch/suse-build-key-*.rpm";
+        assert_script_run "rpm --import /usr/lib/rpm/gnupg/keys/*";
+
+        # just for openQA. it expects the firewall to be enabled in the
+        # application tests later.
+        zypper_call "in SuSEfirewall2";
+        assert_script_run "SuSEfirewall2 on";
+    }
+    # disable the openSUSE repos
+    zypper_call('mr -d -a');
+    # register!
+    add_suseconnect_product("SLES", $version, $arch, "--regcode $reg_code");
+
+    # restore product, bug in 42
+    if (is_leap_base('42')) {
+        script_run("ln -s SLES.prod /etc/products.d/baseproduct");
+    }
+
+    # just for the fun of it here
+    script_run("SUSEConnect --list-extensions");
+    # add the modules needed for installation
+    if (is_leap_base('15')) {
+        my @need_modules = ('sle-module-basesystem', 'sle-module-server-applications', 'sle-module-desktop-applications', 'sle-module-legacy', 'PackageHub');
+        for my $module (@need_modules) {
+            add_suseconnect_product($module, $version, $arch);
```

Comment:
JFYI, you actually can pass only name of the module, version and architecture will be automatically injected by `add_suseconnect_product` based on `/etc/os-release`

Reply:
Yes, I will reserve the version and architecture there. Anyway, thanks.

Reply:
I don't think you followed the comment. Shouldn't you just need to write

```perl
add_suseconnect_product($module)
```

here? Then of course you would not need to get the variables `$version` and `$arch` into scope at all

Reply:
Originally I hope to reserved the parameters of version and arch since no different except call two more parameters. Anyway, I removed them now. Thanks.